metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,399 @@
+        self,
+        command: List[str],
+        env: Optional[Dict[str, str]] = None,
+        cwd: Optional[str] = None,
+    ) -> int:
+        """Run a command asynchronously and return its process ID."""
+
+        command_obj = CommandManager(command, env, cwd)
+        pid = await command_obj.async_run()
+        self.commands[pid] = command_obj
+        return pid
+
+    def get(self, pid: int) -> "CommandManager":
+        """Get the CommandManager object for a given process ID."""
+
+        return self.commands.get(pid, None)
+
+    def cleanup(self) -> None:
+        """Clean up log files for all running subprocesses."""
+
+        for v in self.commands.values():
+            v.cleanup()
+
+
+class CommandManager(object):
+    """A manager for an individual subprocess."""
+
+    def __init__(
+        self,
+        command: List[str],
+        env: Optional[Dict[str, str]] = None,
+        cwd: Optional[str] = None,
+    ):
+        self.command = command
+
+        self.env = env if env is not None else os.environ.copy()
+        self.cwd = cwd if cwd is not None else os.getcwd()
+
+        self.process = None
+        self.run_called: bool = False
+        self.log_files: Dict[str, str] = {}
+
+        signal.signal(signal.SIGINT, self.handle_sigint)
+
+    async def __aenter__(self) -> "CommandManager":
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        self.cleanup()
+
+    def handle_sigint(self, signum, frame):
+        """Handle the SIGINT signal."""
+
+        asyncio.create_task(self.kill())
+
+    async def wait(
+        self, timeout: Optional[float] = None, stream: Optional[str] = None
+    ) -> None:
+        """Wait for the subprocess to finish, optionally with a timeout and optionally streaming its output."""
+
+        if timeout is None:
+            if stream is None:
+                await self.process.wait()
+            else:
+                await self.emit_logs(stream)
+        else:
+            try:
+                if stream is None:
+                    await asyncio.wait_for(self.process.wait(), timeout)
+                else:
+                    await asyncio.wait_for(self.emit_logs(stream), timeout)
+            except asyncio.TimeoutError:
+                command_string = " ".join(self.command)
+                await self.kill()
+                print(
+                    "Timeout: The process (PID %d; command: '%s') did not complete within %s seconds."
+                    % (self.process.pid, command_string, timeout)
+                )
+
+    def run(self):
+        if not self.run_called:
+            self.temp_dir = tempfile.mkdtemp()
+            stdout_logfile = os.path.join(self.temp_dir, "stdout.log")
+            stderr_logfile = os.path.join(self.temp_dir, "stderr.log")
+
+            def stream_to_stdout_and_file(pipe, log_file):
+                with open(log_file, "w") as file:
+                    for line in iter(pipe.readline, ""):
```

Comment:
@madhur-ob let's add a `show_output=True` flag in `run` so you can suppress outputs optionally. Here it outputs both the `stdout` and `stderr` of the subprocess to `stdout` always, which is not always preferable.

let's do
```
if  show_output:
    sys.stdout.write(line)
```
even if `show_output=False`, you can retrieve the complete logs from `result.stdout` and `result.stderr`

Reply:
Thanks, I have updated it...